source/azure_iot_hub_client.c

```diff
@@ -0,0 +1,538 @@
+
+    if( ( xAzureIoTHubClientHandle == NULL ) ||
+        ( pucHostname == NULL ) || ( ulHostnameLength == 0 ) ||
+        ( pucDeviceId == NULL ) || ( ulDeviceIdLength == 0 ) ||
+        ( pucBuffer == NULL ) || ( ulBufferLength == 0 ) ||
+        ( xGetTimeFunction == NULL ) || ( pxTransportInterface == NULL ) )
+    {
+        AZLogError( ( "Failed to initialize: Invalid argument \r\n" ) );
+        ret = eAzureIoTHubClientInvalidArgument;
+    }
+    else if ( ( ulBufferLength < azureiotTOPIC_MAX ) ||
+              ( ulBufferLength < ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) ) )
+    {
+        AZLogError( ( "Failed to initialize: Not enough memory passed \r\n" ) );
+        ret = eAzureIoTHubClientOutOfMemory;
+    }
+    else
+    {
+        memset( ( void * ) xAzureIoTHubClientHandle, 0, sizeof( AzureIoTHubClient_t ) );
+
+        /* Setup scratch buffer to be used by middleware */
+        xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientWorkingBufferLength =
+            ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) > azureiotTOPIC_MAX ? ( azureiotUSERNAME_MAX + azureiotPASSWORD_MAX ) : azureiotTOPIC_MAX;
+        xAzureIoTHubClientHandle->_internal.pucAzureIoTHubClientWorkingBuffer = pucBuffer;
+        pucBuffer += xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientWorkingBufferLength;
+        ulBufferLength -= xAzureIoTHubClientHandle->_internal.ulAzureIoTHubClientWorkingBufferLength;
+
+        /* Initialize Azure IoT Hub Client */
+        az_span hostname_span = az_span_create( ( uint8_t * ) pucHostname, ( int32_t ) ulHostnameLength );
+        az_span device_id_span = az_span_create( ( uint8_t * ) pucDeviceId, ( int32_t ) ulDeviceIdLength );
+
+        options = az_iot_hub_client_options_default();
+
+        if( pxHubClientOptions )
+        {
+            options.module_id = az_span_create( ( uint8_t * ) pxHubClientOptions->pucModuleID, ( int32_t ) pxHubClientOptions->ulModuleIDLength );
+            options.model_id = az_span_create( ( uint8_t * ) pxHubClientOptions->pucModelID, ( int32_t ) pxHubClientOptions->ulModelIDLength );
+            options.user_agent = az_span_create( ( uint8_t * ) pxHubClientOptions->pucUserAgent, ( int32_t ) pxHubClientOptions->ulUserAgentLength );
+        }
+        else
+        {
+            options.user_agent = az_span_create( ( uint8_t * ) azureiothubUSER_AGENT, sizeof( azureiothubUSER_AGENT ) - 1 );
+        }
+
+        if( az_result_failed( az_iot_hub_client_init( &xAzureIoTHubClientHandle->_internal.xAzureIoTHubClientCore,
+                                                      hostname_span, device_id_span, &options ) ) )
+        {
+            AZLogError( ( "Failed to initialize az_iot_hub_client_init \r\n" ) );
+            ret = eAzureIoTHubClientInitFailed;
+        }
+        /* Initialize AzureIoTMQTT library. */
+        else if( ( AzureIoTMQTT_Init( &( xAzureIoTHubClientHandle->_internal.xMQTTContext ), pxTransportInterface,
```

Comment:
We should store AzureIoTMQTTResult_t from this call and on failure output value to log.